MatchError with akka.typed.AskPattern

If the typed ask pattern is used on an ActorRef which refers to a nonexistent actor, the AskPattern implementation throws a MatchError (see below; full code is here: https://github.com/durban/akka-typed-examples/blob/8625c5404d109b30a2a5fa0e85137baa7ed2e41f/src/main/scala/com/example/ask/TypedAsk.scala). The reason seems to be that a destructuring val definition tries to match a 3-tuple, but in this case a 2-tuple is returned.

```
[info] Running com.example.ask.TypedAsk 
[ERROR] [03/04/2016 10:34:10.155] [MySys-akka.actor.default-dispatcher-4] [akka.actor.LocalActorRefProvider(akka://MySys)] guardian failed, shutting down system
scala.MatchError: (Actor[akka://MySys/deadLetters],scala.concurrent.impl.Promise$KeptPromise@1603220d) (of class scala.Tuple2)                                
        at akka.typed.AskPattern$PromiseRef.<init>(Ask.scala:42)                                                                                              
        at akka.typed.AskPattern$PromiseRef$.apply(Ask.scala:60)                                                                                              
        at akka.typed.AskPattern$.ask(Ask.scala:64)                                                                                                           
        at akka.typed.AskPattern$Askable$.$qmark$extension(Ask.scala:38)                                                      
        at com.example.ask.TypedAsk$$anonfun$root$1$$anonfun$apply$2.apply(TypedAsk.scala:26)                                 
        at com.example.ask.TypedAsk$$anonfun$root$1$$anonfun$apply$2.apply(TypedAsk.scala:22)                                 
        at akka.typed.ScalaDSL$Static.message(ScalaDSL.scala:240)                                                             
        at akka.typed.ActorAdapter$$anonfun$receive$1.applyOrElse(Impl.scala:31)                                              
        at akka.actor.Actor$class.aroundReceive(Actor.scala:480)                                                              
        at akka.typed.ActorAdapter.aroundReceive(Impl.scala:17)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
        at akka.actor.ActorCell.invoke(ActorCell.scala:495)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[ERROR] [03/04/2016 10:34:10.160] [MySys-akka.actor.default-dispatcher-4] [akka://MySys/user] (Actor[akka://MySys/deadLetters],scala.concurrent.impl.Promise$KeptPromise@1603220d) (of class scala.Tuple2)
scala.MatchError: (Actor[akka://MySys/deadLetters],scala.concurrent.impl.Promise$KeptPromise@1603220d) (of class scala.Tuple2)
        at akka.typed.AskPattern$PromiseRef.<init>(Ask.scala:42)
        at akka.typed.AskPattern$PromiseRef$.apply(Ask.scala:60)
        at akka.typed.AskPattern$.ask(Ask.scala:64)
        at akka.typed.AskPattern$Askable$.$qmark$extension(Ask.scala:38)
        at com.example.ask.TypedAsk$$anonfun$root$1$$anonfun$apply$2.apply(TypedAsk.scala:26)
        at com.example.ask.TypedAsk$$anonfun$root$1$$anonfun$apply$2.apply(TypedAsk.scala:22)
        at akka.typed.ScalaDSL$Static.message(ScalaDSL.scala:240)
        at akka.typed.ActorAdapter$$anonfun$receive$1.applyOrElse(Impl.scala:31)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:480)
        at akka.typed.ActorAdapter.aroundReceive(Impl.scala:17)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
        at akka.actor.ActorCell.invoke(ActorCell.scala:495)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

^C
```

--- a/akka-typed/src/main/scala/akka/typed/Ask.scala
+++ b/akka-typed/src/main/scala/akka/typed/Ask.scala
@@ -45,7 +45,9 @@ object AskPattern {
   }
 
   private class PromiseRef[U](target: ActorRef[_], untyped: InternalActorRef, timeout: Timeout) {
-    val (ref: ActorRef[U], future: Future[U], promiseRef: PromiseActorRef) =
+
+    // Note: _promiseRef mustn't have a type pattern, since it can be null
+    private[this] val (_ref: ActorRef[U], _future: Future[U], _promiseRef) =
       if (untyped.isTerminated)
         (
           adapter.ActorRefAdapter[U](untyped.provider.deadLetters),
@@ -59,6 +61,10 @@ object AskPattern {
         val b = adapter.ActorRefAdapter[U](a)
         (b, a.result.future.asInstanceOf[Future[U]], a)
       }
+
+    val ref: ActorRef[U] = _ref
+    val future: Future[U] = _future
+    val promiseRef: PromiseActorRef = _promiseRef
   }
 
   private def askUntyped[T, U](target: ActorRef[T], untyped: InternalActorRef, timeout: Timeout, f: ActorRef[U] ⇒ T): Future[U] = {

--- a/akka-typed/src/test/scala/akka/typed/AskSpec.scala
+++ b/akka-typed/src/test/scala/akka/typed/AskSpec.scala
@@ -1,0 +1,82 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+import org.scalatest.concurrent.ScalaFutures
+
+import akka.util.Timeout
+import akka.pattern.AskTimeoutException
+
+import ScalaDSL._
+import AskPattern._
+
+object AskSpec {
+
+  sealed trait Msg
+  final case class Foo(s: String)(val replyTo: ActorRef[String]) extends Msg
+  final case class Stop(replyTo: ActorRef[Unit]) extends Msg
+}
+
+class AskSpec extends TypedSpec with ScalaFutures {
+
+  import AskSpec._
+
+  trait Common {
+
+    def system: ActorSystem[TypedSpec.Command]
+
+    implicit def executor: ExecutionContext =
+      system.executionContext
+
+    val behavior: Behavior[Msg] = Total[Msg] {
+      case foo @ Foo(_) ⇒
+        foo.replyTo ! "foo"
+        Same
+      case Stop(r) ⇒
+        r ! ()
+        Stopped
+    }
+
+    def `must fail the future if the actor is already terminated`(): Unit = {
+      val fut = for {
+        ref ← system ? TypedSpec.Create(behavior, "test1")
+        _ ← ref ? Stop
+        answer ← ref.?(Foo("bar"))(Timeout(1.second), implicitly)
+      } yield answer
+      (fut.recover { case _: AskTimeoutException ⇒ "" }).futureValue should ===("")
+    }
+
+    def `must succeed when the actor is alive`(): Unit = {
+      val fut = for {
+        ref ← system ? TypedSpec.Create(behavior, "test2")
+        answer ← ref ? Foo("bar")
+      } yield answer
+      fut.futureValue should ===("foo")
+    }
+  }
+
+  object `Ask pattern (native)` extends Common with NativeSystem
+
+  object `Ask pattern (adapted)` extends Common with AdaptedSystem {
+
+    import AskSpec._
+
+    /** See issue #19947 (MatchError with adapted ActorRef) */
+    def `must fail the future if the actor doesn't exist`(): Unit = {
+      val noSuchActor: ActorRef[Msg] = system match {
+        case adaptedSys: adapter.ActorSystemAdapter[_] ⇒
+          adapter.actorRefAdapter(adaptedSys.untyped.provider.resolveActorRef("/foo/bar"))
+        case _ ⇒
+          fail("this test must only run in an adapted actor system")
+      }
+      val fut = for {
+        answer ← noSuchActor.?(Foo("bar"))(Timeout(1.second), implicitly)
+      } yield answer
+      (fut.recover { case _: AskTimeoutException ⇒ "" }).futureValue should ===("")
+    }
+  }
+}

